crates/ethrpc/src/current_block/mod.rs

```diff
@@ -16,6 +16,130 @@ use {
     },
 };
 
+/// Internals of [`CurrentBlockStream`] that allow for only watching the most
+/// recent block as well as all blocks with the same handle.
+#[derive(Debug)]
+struct BlockStreamInner {
+    watch_receiver: watch::Receiver<BlockInfo>,
+    watch_sender: watch::Sender<BlockInfo>,
+    broadcast_sender: broadcast::Sender<BlockInfo>,
+}
+
+impl BlockStreamInner {
+    fn publish_block(&self, value: BlockInfo) {
+        let _ = self.broadcast_sender.send(value);
+        let _ = self.watch_sender.send(value);
+    }
+
+    fn new(initial_block: BlockInfo) -> Self {
+        /// This many blocks may be buffered in any [`broadcast::Receiver`].
+        const MAX_SIZE: usize = 1_000;
+
+        let (watch_sender, watch_receiver) = watch::channel(initial_block);
+        let (broadcast_sender, _broadcast_receiver) = broadcast::channel(MAX_SIZE);
+
+        Self {
+            watch_receiver,
+            watch_sender,
+            broadcast_sender,
+        }
+    }
+}
+
+/// Monitors the blockchain for new blocks and allows users to get notified
+/// about them. Can be considered the "heartbeat" of the backend.
+#[derive(Debug, Clone)]
+pub struct CurrentBlockStream(Arc<BlockStreamInner>);
+
+impl CurrentBlockStream {
+    /// Returns a reference to the current block. Keeping this reference
+    /// alive locks a mutex so it should be used as shortly as possible.
+    pub fn current(&self) -> watch::Ref<'_, BlockInfo> {
+        self.0.watch_receiver.borrow()
+    }
+
+    /// Returns a stream that always yields the most recent block (if unseen).
+    /// Also yields the current block right away and not just the next new
+    /// block. Use this if you only need to know about the current state.
+    pub fn watch_stream(&self) -> WatchStream<BlockInfo> {
+        WatchStream::new(self.0.watch_receiver.clone())
+    }
+
+    /// Returns a stream that yields every new block **AFTER** it got created.
+    /// Use this if it's required to not miss any blocks.
+    pub fn buffering_stream(&self) -> impl Stream<Item = BlockInfo> {
```

Comment:
This is bound by `MAX_SIZE` right? Let's add that as a comment.

Also we should comment on how the buffer deals with reorgs (it seems it doesn't)

Reply:
Also, where is this used now?

Reply:
After discovering that the indexing logic actually does not need to see every single block to work correctly as I initially thought this function did not end up being used. So technically this PR could get simplified a bit again although I don't really see harm in keeping it now that we have it.